src/providers/pkcs11_provider/key_metadata.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use super::LocalIdStore;
+use super::{KeyInfo, Pkcs11Provider};
+use crate::key_info_managers;
+use crate::key_info_managers::{KeyTriple, ManageKeyInfo};
+use log::{error, warn};
+use parsec_interface::operations::psa_key_attributes::*;
+use parsec_interface::requests::{ResponseStatus, Result};
+use std::sync::RwLock;
+
+impl Pkcs11Provider {
+    // This function returns the `RWLocks` found on the `Pkcs11Provider`
+    // in the order in which they should *always* be taken. Changing the order
+    // of locking in one method can very easily result in deadlocking.
+    pub(super) fn get_ordered_locks<'a>(
+        &'a self,
+    ) -> (
+        &'a RwLock<dyn ManageKeyInfo + Send + Sync>,
+        &'a RwLock<LocalIdStore>,
+    ) {
+        (&self.key_info_store, &self.local_ids)
+    }
+
+    /// Gets a key identifier and key attributes from the Key Info Manager.
+    pub(super) fn get_key_info(&self, key_triple: &KeyTriple) -> Result<([u8; 4], Attributes)> {
+        let locks = self.get_ordered_locks();
+        let store_handle = locks.0.read().expect("Local ID lock poisoned");
+        match store_handle.get(key_triple) {
+            Ok(Some(key_info)) => {
+                if key_info.id.len() == 4 {
+                    let mut dst = [0; 4];
+                    dst.copy_from_slice(&key_info.id);
+                    Ok((dst, key_info.attributes))
+                } else {
+                    error!("Stored Key ID is not valid.");
+                    Err(ResponseStatus::KeyInfoManagerError)
+                }
+            }
+            Ok(None) => Err(ResponseStatus::PsaErrorDoesNotExist),
+            Err(string) => Err(key_info_managers::to_response_status(string)),
+        }
+    }
+
+    pub(super) fn create_key_id(
+        &self,
+        key_triple: KeyTriple,
+        key_attributes: Attributes,
+    ) -> Result<[u8; 4]> {
+        let locks = self.get_ordered_locks();
+        let mut store_handle = locks.0.write().expect("Key store lock poisoned");
+        let mut local_ids_handle = locks.1.write().expect("Local ID lock poisoned");
+        let mut key_id = rand::random::<[u8; 4]>();
+        while local_ids_handle.contains(&key_id) {
+            key_id = rand::random::<[u8; 4]>();
+        }
```

Comment:
I don't know if that is not over-optimising but you could modify the code and create new blocks so that each locks is taken the least amount of time needed.
Like:
```diff
-match store_handle.insert(key_triple.clone(), key_info) {
+match {
+    let mut store_handle = locks.0.write().expect("Key store lock poisoned");
+    store_handle.insert(key_triple.clone(), key_info)
+} {
```
and something similar for the `locks.1`

Maybe that is over the top though and adds more complexity than needed. What do you think?

Reply:
Ok, discussed this but given the issues we've encountered with multithreading and SoftHSM2, we're adding a mutex around all PKCS11 operations, so this wouldn't be that helpful. Plus, it'd probably fall under "premature optimization"